pkg/filestore/s3/s3.go

```diff
@@ -0,0 +1,102 @@
+package s3
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"scratchdata/config"
+	"scratchdata/pkg/filestore"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+// Storage implements filestore.StorageBackend using Amazon S3
+type Storage struct {
+	ctx        context.Context
+	client     *s3.S3
+	downloader *s3manager.Downloader
+	accessKey  string
+	bucket     string
+}
+
+// Upload implements filestore.StorageBackend.Upload
+func (s *Storage) Upload(path string, r io.ReadSeeker) error {
+	// keep track of the file position in-case we need to retry the upload
+	rPos, err := r.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return fmt.Errorf("Storage.Upload: Cannnot get current file position: %w", err)
+	}
+
+	input := &s3.PutObjectInput{
+		Bucket:             aws.String(s.bucket),
+		Key:                aws.String(path),
+		Body:               r,
+		ContentDisposition: aws.String("attachment"),
+	}
+
+	_, err = s.client.PutObjectWithContext(s.ctx, input)
+	if err == nil {
+		return nil
+	}
+	var awsErr awserr.Error
+	if !errors.As(err, &awsErr) || awsErr.Code() != s3.ErrCodeNoSuchBucket {
+		return fmt.Errorf("Storage.Upload: %s: %w", path, err)
+	}
+
+	// if the bucket doesn't exist, create it and try again
```

Comment:
I like the thought, but let's omit auto-creating buckets for now.

Reply:
done